.github/workflows/tests.yml

```diff
@@ -31,29 +31,43 @@ jobs:
           - ubuntu-latest
           - macos-latest
           - windows-latest
-        python-version:
-          - "3.7"
-          - "3.8"
-          - "3.9"
-          - "3.10"
-          - "3.11"
-          - "pypy-3.7"
 
     steps:
       - name: "Check out the repo"
         uses: "actions/checkout@v3"
 
       - name: "Set up Python"
+        id: "setup-python"
         uses: "actions/setup-python@v4"
         with:
-          python-version: "${{ matrix.python-version }}"
-          cache: pip
-          cache-dependency-path: 'requirements/*.txt'
+          # The last listed Python version is the default.
+          python-version: |
+            pypy-3.7
+            3.7
+            3.8
+            3.9
+            3.10
+            3.11
+
+      - name: "Restore cache"
+        id: "restore-cache"
+        uses: "actions/cache@v3"
+        with:
+          path: |
+            .tox/
+            .venv/
+          key: "cache-python-${{ steps.setup-python.outputs.python-version }}-os-${{ runner.os }}-hash-${{ hashFiles('tox.ini', 'requirements/*.txt') }}"
+
+      - name: "Identify venv path"
+        shell: "bash"
+        run: "echo 'venv-path=${{ runner.os == 'Windows' && '.venv/Scripts' || '.venv/bin' }}' >> $GITHUB_ENV"
 
       - name: "Install dependencies"
+        if: "steps.restore-cache.outputs.cache-hit == false"
         run: |
-          python -m pip install -U setuptools
-          python -m pip install -r requirements/tox.txt
+          python -m venv .venv
+          ${{ env.venv-path }}/python -m pip install -U setuptools
+          ${{ env.venv-path }}/python -m pip install -r requirements/tox.txt
```

Comment:
This step is expensive, so it only happens when there's a cache miss. A virtual environment is used because it's far easier to cache a virtual environment at a path of our choosing than to try caching items in an OS-specific location.